### PR TITLE
core/vm, protocol_params: implement eip-2565 modexp repricing

### PR DESCRIPTION
**Pull request by holiman**
This PR implements [EIP-2565](https://eips.ethereum.org/EIPS/eip-2565), at least in the latest (maybe still unmerged version). 

Screenshot_2020-09-24 Untitled spreadsheet

Running this on my local machine, the gas-costs are within reasonable bounds, compared to other precompiles, with the worst case being `nagydani-1-pow0x10001-Gas`. In general, the variance is a lot smaller, with not as many wildly overpriced outliers. 

This PR only implements the EIP, but does not in any way activate it nor make it 'activateable', which would be a future PR. However, if we want the possibility of doing benchmarking and e.g. combine this EIP with other changes to call costs, it might be nice to have it in.  But the intent is that it should be safe to merge this PR.

### CODE REVIEW

File: core/vm/contracts.go [hunk 1]
```diff
@@ -85,7 +85,7 @@ var PrecompiledContractsYoloV1 = map[common.Address]PrecompiledContract{
 	common.BytesToAddress([]byte{2}):  &sha256hash{},
 	common.BytesToAddress([]byte{3}):  &ripemd160hash{},
 	common.BytesToAddress([]byte{4}):  &dataCopy{},
-	common.BytesToAddress([]byte{5}):  &bigModExp{},
+	common.BytesToAddress([]byte{5}):  &bigModExp{false},
```
Comment (karalabe) [inline core/vm/contracts.go:88]: Nits, please add the field name, otherwise `{false}` is kind of hard to parse.

File: core/vm/contracts.go [hunk 2]
```diff
@@ -217,6 +220,48 @@ var (
...
 
+// multComplexity implements bigModexp multComplexity formula, as defined in EIP-198
+//
+// def mult_complexity(x):
+//    if x <= 64: return x ** 2
+//    elif x <= 1024: return x ** 2 // 4 + 96 * x - 3072
+//    else: return x ** 2 // 16 + 480 * x - 199680
+//
+// where is x is max(length_of_MODULUS, length_of_BASE)
+func multComplexity(x *big.Int) *big.Int {
```
Comment (karalabe) [inline core/vm/contracts.go:231]: Please give a more meaningful name to this method. `vm.multComplexity` seems dangling. Maybe add ModExp somewhere or maybe we could make it a method on `bigModExp`?

File: core/vm/contracts.go [hunk 3]
```diff
@@ -251,23 +296,25 @@ func (c *bigModExp) RequiredGas(input []byte) uint64 {
...
-			new(big.Int).Div(new(big.Int).Mul(gas, gas), big16),
-			new(big.Int).Sub(new(big.Int).Mul(big480, gas), big199680),
-		)
+	if c.eip2565 {
+		// EIP-2565 has three changes
+		// 1. Different multComplexity
+		gas = multComplexityEIP2565(gas)
+		gas.Mul(gas, math.BigMax(adjExpLen, big1))
+		// 2. Different divisor (`GQUADDIVISOR`)
+		gas.Div(gas, new(big.Int).SetUint64(params.ModExpQuadCoeffDivEIP2565))
```
Comment (karalabe) [inline core/vm/contracts.go:307]: Might as well make this into a constant if we have big1, big2, big199680 etc.

File: core/vm/contracts.go [hunk 4]
```diff
@@ -251,23 +296,25 @@ func (c *bigModExp) RequiredGas(input []byte) uint64 {
 		adjExpLen.Mul(big8, adjExpLen)
 	}
 	adjExpLen.Add(adjExpLen, big.NewInt(int64(msb)))
```
Comment (karalabe) [inline core/vm/contracts.go:298]: The new EIP at least has some funky part:

```
def calculate_iteration_count(exponent_length, exponent):
    iteration_count = 0
    if exponent_length <= 32 and exponent == 0: iteration_count = 0
    elif exponent_length <= 32: iteration_count = exponent.bit_length() - 1
    elif exponent_length > 32: iteration_count = (8 * (exponent_length - 32)) + ((exponent & (2**256 - 1)).bit_length() - 1)
    return max(iteration_count, 1)
```

Namely, in the 3rd clause, it has `((exponent & (2**256 - 1)).bit_length()` vs `exponent.bit_length()` in the second. The code however treats both cases as `msb`. Is this correct?

Alternatively, `& (2**256 - 1)` seems to mean "truncate to 32 bytes". But when can it be larger than 32 bytes? Can it? If not, I think we need to clear up the spec as this is super weird.

File: core/vm/contracts.go [hunk 5]
```diff
@@ -251,23 +296,25 @@ func (c *bigModExp) RequiredGas(input []byte) uint64 {
...
-		)
-	default:
-		gas = new(big.Int).Add(
-			new(big.Int).Div(new(big.Int).Mul(gas, gas), big16),
-			new(big.Int).Sub(new(big.Int).Mul(big480, gas), big199680),
-		)
+	if c.eip2565 {
+		// EIP-2565 has three changes
+		// 1. Different multComplexity
+		gas = multComplexityEIP2565(gas)
```
Comment (karalabe) [inline core/vm/contracts.go:304]: In all honesty I'm not sure it's worthwhile to keep this method separated, since it's just 3 lines and since the rest of the gas calculation is inclined anyway. IMHO let's just inline those 3 lines too.

File: core/vm/contracts.go (hunk 4 above)
Comment (holiman) [reply core/vm/contracts.go:298]: I think it means truncate. I'm not sure -- that is old code..

Comment (holiman) [reply core/vm/contracts.go:286]: I hadn't seen that final version of the EIP -- however, the EIP consists of three separate changes to the previous version. That's what's implemented.

File: core/vm/contracts.go [hunk 6]
```diff
@@ -58,7 +58,7 @@ var PrecompiledContractsByzantium = map[common.Address]PrecompiledContract{
 	common.BytesToAddress([]byte{2}): &sha256hash{},
 	common.BytesToAddress([]byte{3}): &ripemd160hash{},
 	common.BytesToAddress([]byte{4}): &dataCopy{},
-	common.BytesToAddress([]byte{5}): &bigModExp{},
+	common.BytesToAddress([]byte{5}): &bigModExp{false},
```
Comment (karalabe) [inline core/vm/contracts.go:61]: ;)

Comment (holiman) [reply core/vm/contracts.go]: fixed!

**Review by karalabe — APPROVED** [review]
LGTM
